src/Middleware/OutputCaching/src/CacheVaryByRules.cs

```diff
@@ -20,10 +20,15 @@ public sealed class CacheVaryByRules
     /// </summary>
     public IDictionary<string, string> VaryByCustom => _varyByCustom ??= new();
 
+    /// <summary>
+    /// Gets or sets the list of route values to vary by.
+    /// </summary>
+    public StringValues RouteValues { get; set; }
+
     /// <summary>
     /// Gets or sets the list of headers to vary by.
     /// </summary>
-    public StringValues Headers { get; set; }
+    public StringValues HeaderNames { get; set; }
 
     /// <summary>
     /// Gets or sets the list of query string keys to vary by.
```

Comment:
This is below the cutoff for review comments, but here's the doc comment for `VaryByPrefix`:

```csharp
    /// <summary>
    /// Gets or sets a prefix to vary by.
    /// </summary>
    public StringValues VaryByPrefix { get; set; }
```

Get's or sets a prefix to what exactly? We should spell this out in the doc comment.

Reply:
It's a prefix for the cache key. But after adding some tests this week I feel like this `StringValues` is wrong. In ResponseCaching it's `string VaryByKeyPrefix` and it makes more sense. I assume I changed it to `StringValues` so I can put the custom values in it as a backing field when normalized, but in the end it is confusing.

I think we should change it by making it a `string VaryByKeyPrefix`, like in Response Caching, to add some kind of discriminator to the cache key (tenant key for instance). And then isolate it from `VaryByCustom`. The `VaryByValue()` call would also always take a key. Right now when a key is not set it adds it to the `VaryByPrefix`.